modules/dnn/perf/perf_layer.cpp

```diff
@@ -739,6 +739,62 @@ PERF_TEST_P_(Layer_InstanceNorm, InstanceNorm)
     test_layer({N, C, H, W});
 }
 
+struct Layer_Attention : public TestBaseWithParam<tuple<Backend, Target>> {
+    void test_layer(const std::vector<int> x_shape, const std::vector<int> qkv_hidden_sizes, const int num_heads) {
+        int backendId = get<0>(GetParam());
+        int targetId = get<1>(GetParam());
+
+        auto qk_hidden_size = qkv_hidden_sizes[0];
+        auto v_hidden_size = qkv_hidden_sizes[2];
+
+        auto input_hidden_size = x_shape[2];
+        auto hidden_size = qk_hidden_size + qk_hidden_size + v_hidden_size;
+
+        Mat x(x_shape, CV_32F);
+        Mat weight(std::vector<int>{input_hidden_size, hidden_size}, CV_32F);
+        Mat bias(std::vector<int>{hidden_size}, CV_32F);
+
+        randu(x, 0.f, 1.f);
+        randu(weight, 0.f, 1.f);
+        randu(bias, 0.f, 1.f);
+
+        LayerParams lp;
+        lp.type = "Attention";
```

Comment:
How the numbers from https://github.com/opencv/opencv/pull/24476#issue-1971965496 can be reproduced? I mean does it make sense to run test on an ONNX model to catch the fusion too?

Reply:
> I mean does it make sense to run test on an ONNX model to catch the fusion too?

Yes, definitely. I will add a test case for attention fusion in `test_graph_simplifier`.

> How the numbers from https://github.com/opencv/opencv/pull/24476#issue-1971965496 can be reproduced?

For `w/o Attention`, I use the onnx model and test it in `perf_layer.cpp` **with `AttentionSubgraph` disabled**:

```c++
struct Layer_Attention_expanded : public TestBaseWithParam<tuple<Backend, Target>> {
    void test_layer(const std::vector<int> x_shape, const std::vector<int> qkv_hidden_sizes, const int num_heads) {
        int backendId = get<0>(GetParam());
        int targetId = get<1>(GetParam());

        Mat x(x_shape, CV_32F);
        randu(x, 0.f, 1.f);

        auto net = readNet("/path/to/attention.onnx");

        {
            net.setInput(x);

            net.setPreferableBackend(backendId);
            net.setPreferableTarget(targetId);
            Mat out = net.forward();
        }

        TEST_CYCLE()
        {
            Mat out = net.forward();
        }

        SANITY_CHECK_NOTHING();
    }
};

PERF_TEST_P_(Layer_Attention_expanded, VisionTransformer) {
    test_layer({1, 197, 768}, {768, 768, 768}, 12);
}

INSTANTIATE_TEST_CASE_P(/**/, Layer_Attention_expanded, testing::Values(std::make_tuple(DNN_BACKEND_OPENCV, DNN_TARGET_CPU)));
```

This code is not pushed.

Also python script to run benchmark with ORT:

```python
import time

import numpy as np
import onnxruntime as ort

np.random.seed(0)

x = np.random.rand(1, 197, 768).astype(np.float32)

net = ort.InferenceSession("./attention_subgraph.onnx", providers=["CPUExecutionProvider"])

# warmup
net.run([], {"x": x})

# benchmark
all_times = []
for _ in range(100):
    start = time.time()
    net.run([], {"x": x})
    end = time.time()
    all_times.append(1000 * (end - start))

all_times = sorted(all_times)
mean = sum(all_times) / len(all_times)
median = (all_times[49] + all_times[50]) / 2
minimum = min(all_times)

print("mean={:.2f}, median={:.2f}, min={:.2f}".format(mean, median, minimum))
```

Reply:
Thanks! Sure, That's ok, I just wanted to keep aligned in backends benchmarks.